libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp

```diff
@@ -613,54 +639,72 @@ void  NavEKF3_core::updateFilterStatus(void)
 {
-    if (yawEstimator != nullptr && frontend->_fusionModeGPS <= 1) {
-        float trueAirspeed;
-        if (is_positive(defaultAirSpeed) && assume_zero_sideslip()) {
-            if (imuDataDelayed.time_ms - tasDataDelayed.time_ms < 5000) {
-                trueAirspeed = tasDataDelayed.tas;
-            } else {
-                trueAirspeed = defaultAirSpeed * dal.get_EAS2TAS();
-            }
+    // exit immediately if no yaw estimator
+    if (yawEstimator == nullptr) {
+        return;
+    }
+
+    // ensure GPS is used for horizontal position and velocity
+    if (frontend->_sources.getPosXYSource() != AP_NavEKF_Source::SourceXY::GPS ||
+        frontend->_sources.getVelXYSource() != AP_NavEKF_Source::SourceXY::GPS) {
+        return;
+    }
+
+    float trueAirspeed;
+    if (is_positive(defaultAirSpeed) && assume_zero_sideslip()) {
+        if (imuDataDelayed.time_ms - tasDataDelayed.time_ms < 5000) {
+            trueAirspeed = tasDataDelayed.tas;
         } else {
-            trueAirspeed = 0.0f;
+            trueAirspeed = defaultAirSpeed * dal.get_EAS2TAS();
         }
-
-        yawEstimator->update(imuDataDelayed.delAng, imuDataDelayed.delVel, imuDataDelayed.delAngDT, imuDataDelayed.delVelDT, EKFGSF_run_filterbank, trueAirspeed);
+    } else {
+        trueAirspeed = 0.0f;
     }
+
+    yawEstimator->update(imuDataDelayed.delAng, imuDataDelayed.delVel, imuDataDelayed.delAngDT, imuDataDelayed.delVelDT, EKFGSF_run_filterbank, trueAirspeed);
 }
 
 void NavEKF3_core::runYawEstimatorCorrection()
 {
-    if (yawEstimator != nullptr && frontend->_fusionModeGPS <= 1) {
-        if (gpsDataToFuse) {
-            Vector2f gpsVelNE = Vector2f(gpsDataDelayed.vel.x, gpsDataDelayed.vel.y);
-            float gpsVelAcc = fmaxf(gpsSpdAccuracy, frontend->_gpsHorizVelNoise);
-            yawEstimator->fuseVelData(gpsVelNE, gpsVelAcc);
-
-            // after velocity data has been fused the yaw variance esitmate will have been refreshed and
-            // is used maintain a history of validity
-            float yawEKFGSF, yawVarianceEKFGSF;
-            bool canUseEKFGSF = yawEstimator->getYawData(yawEKFGSF, yawVarianceEKFGSF) && is_positive(yawVarianceEKFGSF) && yawVarianceEKFGSF < sq(radians(GSF_YAW_ACCURACY_THRESHOLD_DEG));
-            if (canUseEKFGSF) {
-                if (EKFGSF_yaw_valid_count <  GSF_YAW_VALID_HISTORY_THRESHOLD) {
-                    EKFGSF_yaw_valid_count++;
-                }
-            } else {
-                EKFGSF_yaw_valid_count = 0;
+    // exit immediately if no yaw estimator
+    if (yawEstimator == nullptr) {
+        return;
+    }
+
+    // ensure GPS is used for horizontal position and velocity
+    if (frontend->_sources.getPosXYSource() != AP_NavEKF_Source::SourceXY::GPS ||
+        frontend->_sources.getVelXYSource() != AP_NavEKF_Source::SourceXY::GPS) {
+        return;
+    }
+
+    if (gpsDataToFuse) {
+        Vector2f gpsVelNE = Vector2f(gpsDataDelayed.vel.x, gpsDataDelayed.vel.y);
```

Comment:
I'm sure this is an improvement but this is existing code that has simply been moved.  I'm going to resist changing this because it could lead to calls for more testing and slow the merge.